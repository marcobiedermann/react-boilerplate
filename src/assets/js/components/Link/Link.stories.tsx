import { ComponentMeta, ComponentStoryObj } from '@storybook/react';
import Link from '.';

export default {
  component: Link,
} as ComponentMeta<typeof Link>;

export const Default: ComponentStoryObj<typeof Link> = {
  args: {
    children: 'Link',
    to: 'http://example.com',
  },
};

export const Internal: ComponentStoryObj<typeof Link> = {
  args: {
    children: 'Link',
    to: '/',
  },
};
