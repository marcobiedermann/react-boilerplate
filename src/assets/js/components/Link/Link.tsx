import { Link as ReactRouterLink, LinkProps } from 'react-router-dom';

type LinkProps = LinkProps;

function Link(props: LinkProps): JSX.Element {
  const { to, ...otherProps } = props;
  const isInternal = /^\/(?!\/)/.test(to);

  if (isInternal) {
    return <ReactRouterLink to={to} {...otherProps} />;
  }

  return <a href={to} target="_blank" rel="noopener noreferrer" {...otherProps} />;
}

export type { LinkProps };
export default Link;
