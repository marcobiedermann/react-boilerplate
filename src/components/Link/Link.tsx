import { Link as ReactRouterLink, LinkProps as ReactRouterLinkProps } from 'react-router-dom';

interface LinkProps extends ReactRouterLinkProps {
  to: string;
}

function Link(props: LinkProps): JSX.Element {
  const { children, to, ...otherProps } = props;
  const isInternal = /^\/(?!\/)/.test(to);

  if (isInternal) {
    return (
      <ReactRouterLink to={to} {...otherProps}>
        {children}
      </ReactRouterLink>
    );
  }

  return (
    <a href={to} target="_blank" rel="noopener noreferrer" {...otherProps}>
      {children}
    </a>
  );
}

export type { LinkProps };
export default Link;
