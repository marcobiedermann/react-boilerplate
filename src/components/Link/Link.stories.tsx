import type { Meta, StoryObj } from '@storybook/react';
import Link from '.';

const meta: Meta<typeof Link> = {
  component: Link,
};

type Story = StoryObj<typeof Link>;

const Default: Story = {
  args: {
    children: 'Link',
    to: 'http://example.com',
  },
};

const Internal: Story = {
  args: {
    children: 'Link',
    to: '/',
  },
};

export { Default, Internal };
export default meta;
