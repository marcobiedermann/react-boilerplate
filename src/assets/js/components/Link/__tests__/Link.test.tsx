import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, expect, it } from 'vitest';
import * as stories from '../Link.stories';

const { Default, Internal } = composeStories(stories);

describe('link', () => {
  it('should render correctly', () => {
    const { container } = render(<Default />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render internal correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <Internal />
      </MemoryRouter>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
