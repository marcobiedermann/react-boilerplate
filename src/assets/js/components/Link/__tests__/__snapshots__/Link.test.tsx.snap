// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link should render correctly 1`] = `
<a
  href="http://example.com"
  rel="noopener noreferrer"
  target="_blank"
>
  Link
</a>
`;

exports[`link should render internal correctly 1`] = `
<a
  href="/"
>
  Link
</a>
`;
